Add openid connect quickstart test

cemnura opened the pull request:
Fixes #355

**Check list**:

Your pull request:

- [x] targets the `development` branch
- [x] uses the `999-SNAPSHOT` version of Quarkus
- [x] has tests (`mvn clean test`)
- [x] works in native (`mvn clean package -Pnative`)
- [x] has native tests (`mvn clean verify -Pnative`)
- [x] makes sure the documentation must not be updated
- [ ] links the documentation update pull request (if needed)
- [ ] updates or creates the `README.md` file (with build and run instructions)


I would also enjoy improving the documentation for oauth2 mentioned in  #521

cescoffier commented on security-oauth2-quickstart/src/main/resources/application.properties:
```diff
@@ -1,3 +1,4 @@
 quarkus.oauth2.client-secret=secret
-quarkus.oauth2.introspection-url=http://oauth-server/introspect
\ No newline at end of file
+quarkus.oauth2.introspection-url=http://localhost:8180/auth/realms/quarkus/protocol/openid-connect/token/introspect
```
Wondering if this property can use a random port (and set by a Quarkus Test Resource)

cemnura replied:
This configuration is overwritten during at QuarkusTestResource creation. I was not able to delete this configuration because the following error occurs if not written.
```
io.quarkus.runtime.configuration.ConfigurationException: client-id, client-secret and introspection-url must be configured when the oauth2 extension is enabled
```

cescoffier commented on security-oauth2-quickstart/src/test/java/org/acme/security/oauth2/TokenSecuredResourceTest.java:
```diff
@@ -0,0 +1,76 @@
+@QuarkusTest
+public class TokenSecuredResourceTest {
+
+    private static final String KEYCLOAK_SERVER_URL = System.getProperty("keycloak.url", "http://localhost:8180/auth");
```
So we cannot use TestContainer like this, as the lifecycle is different from the Quarkus lifecycle.

You need to use a Quarkus Test Resource, and BTW, it will also let you set some properties dynamically. 
Check how the Kafka quickstart is built as an example.

cemnura replied:
I will check how the Kafka quickstart is built.

cescoffier commented on security-oauth2-quickstart/src/test/java/org/acme/security/oauth2/TokenSecuredResourceTest.java:
```diff
@@ -0,0 +1,76 @@
+
+    @ClassRule
+    @Container
+    private static GenericContainer keycloak = new FixedHostPortGenericContainer("quay.io/keycloak/keycloak:9.0.3")
```
This should be moved to a Quarkus Test Resources, and probably bound to a free port. 

Also, I'm surprised about @ClassRule as it's a Junit 4 annotation.

cemnura replied:
Okey I will move try to move this and try to bound to a free port.

Yes you are correct ``@ClassRule`` is a Junit4 annotation. I misread the testcontainer documentation. I will be removing this in general.

cescoffier commented on security-oauth2-quickstart/src/test/resources/.notes:
```diff
@@ -0,0 +1,66 @@
+curl -X POST \
```
What the goal of this file?

cemnura replied:
These where my notes when developing this qucikstart. I thought that I may create documentation using these commands.

cemnura replied:
I will delete these files when I am going to squash commits.

cescoffier reviewed (dismissed):
We cannot use the  `@TestContainer` as it introduces a lifecycle mismatch. You need to define a Quarkus Test Resource (and use the TestContainer API in it).

cescoffier commented:
@pedroigor @sberyozkin  can you review the "security" part? There is an endpoint to retrieve tokens. While I understand the usage in tests, it should be mentioned somewhere to not use this in production (right?)

cemnura commented:
> We cannot use the `@TestContainer` as it introduces a lifecycle mismatch. You need to define a Quarkus Test Resource (and use the TestContainer API in it).

I have defined a Quarkus Test Resource to enable the creation of a keycloak server for testing.

sberyozkin commented:
@cescoffier the `oauth2` extension was added by Loic originally to support the opaque (binary) tokens, they should be available to the service code as `SecurityIdentity.getCredential(TokenCredential.class)` - it is probably up to the application what to do with this token if they want to propagate.
@cemnura `quarkus-oidc`, `service` application type - JWT will be verified by the locally cached JWK keys while binary tokens should be introspected (though we don't have a test for the binary token case)

cemnura commented:
Hello @sberyozkin, I did not completely understand you. Do you think there are missing test?

loicmathieu commented:
@cemnura just saw your PR after I created #609 to add a test and a mock of the authentication server via WireMock.

Reading your PR, it appears that this is an OIDC based test, it should be better to add it to the `security-openid-connect-quickstart` quickstart that also miss some test as the OAuth2 extension is a lightweight security implementation for OAuth2 opaque token (no JWT, no OIDC).

pedroigor commented:
@loicmathieu I agree. It seems that the tests herein are basically checking whether or not the extension is capable of also introspecting JWT tokens if comparing to yours. Although still a valid OAuth2 usage (you may have authorization servers issuing JWTs too). Your tests seem to be simple as they don't require a Keycloak container running. One thing I'm missing though is checks for failures when the introspection endpoint says that the token is not active.

And yeah, a similar test to oidc quickstart is necessary.

cemnura commented:
Hello @loicmathieu thanks for viewing this PR. Would you think this unit test would be better in a different quickstart?

@pedroigor I aimed to test the quickstart by testing oauth2 issued tokens from keycloak and send them to the rest api via "Authorization: Bearer" header to validate RBAC. 

The roles for the users are in the  quarkus-realm.json which is a keycloak realm import file.

loicmathieu commented:
@cemnura yes, for me it's a valid integration test for the OIDC quickstart, so moving your code to `security-openid-connect-quickstart` seems a good idea.

@sberyozkin is the guardian of OIDC stuff, so better check with him ;)
I think he said that the same Wiremock setup could be used also for OIDC instead of using a real keycloack server. 
So maybe this would be abandonned, sorry for that ...

Anyway, thanks for trying to improve our quickstart ;)

cemnura commented:
@loicmathieu no problem at all the I believe the Wiremock approach would be a lighter test then starting a keycloak server with docker also making the test dependent to docker.

I will checkout Wiremock for myself as well thanks for the information 🙂 

@sberyozkin I would happily move/improve this integration test to `security-openid-connect-quickstart` as @loicmathieu stated if you see fit.

cemnura commented:
@sberyozkin I have converted the test to the `security-openid-connect-quickstart` quickstart project using TestContainer approach.

I added a Keycloak mock server using WireMock for evaluation and discussion. This still needs work since the server still does not respond to the `protocol/openid-connect/token`

sberyozkin commented:
Hi @cemnura Thanks, it looks good, but I don't think the wiremock code (as well as the docker profile) should be in this PR at all. Please remove it.
Can I suggest, the following, copy and paste `integration-tests/oidc` to `integration-tests/oidc-wiremock` in the main Quarkus respository (and update `integration-tests/pom.xml`) and in this new more module, in its pom, remove the docker and keycloak profiles, remove `KeycloakTestManager` (which sets up keycloak realms/clients) and have this test working with WireMock only

cemnura reviewed (commented):
> Hi @cemnura Thanks, it looks good, but I don't think the wiremock code (as well as the docker profile) should be in this PR at all. Please remove it.
> Can I suggest, the following, copy and paste `integration-tests/oidc` to `integration-tests/oidc-wiremock` in the main Quarkus respository (and update `integration-tests/pom.xml`) and in this new more module, in its pom, remove the docker and keycloak profiles, remove `KeycloakTestManager` (which sets up keycloak realms/clients without Wiremoce) and have this test working with WireMock only (as you have shown here)



> Hi @cemnura Thanks, it looks good, but I don't think the wiremock code (as well as the docker profile) should be in this PR at all. Please remove it.
> Can I suggest, the following, copy and paste `integration-tests/oidc` to `integration-tests/oidc-wiremock` in the main Quarkus respository (and update `integration-tests/pom.xml`) and in this new more module, in its pom, remove the docker and keycloak profiles, remove `KeycloakTestManager` (which sets up keycloak realms/clients without Wiremoce) and have this test working with WireMock only (as you have shown here)

Okey I am updating the PR by removing the WireMock code then migrating it to the new `integration-tests/oidc-wiremock`
as you have recommended.

Will reference this PR in the main Quarkus repository fir refrerence.

cemnura commented on security-openid-connect-quickstart/src/test/java/org/acme/security/openid/connect/KeycloakTestResource.java:
```diff
@@ -0,0 +1,92 @@
+import static com.github.tomakehurst.wiremock.client.WireMock.urlEqualTo;
+import static com.github.tomakehurst.wiremock.core.WireMockConfiguration.options;
+
+public class KeycloakTestResource implements QuarkusTestResourceLifecycleManager {
```
I have deleted this class from this PR and moving it to the main quarkus repository

sberyozkin commented:
@cemnura Can you please rebase and force-push again, as I've just removed `WIP:` in the PR name.

sberyozkin commented on security-openid-connect-quickstart/src/test/java/org/acme/security/openid/connect/UsersResourceTest.java:
```diff
@@ -0,0 +1,43 @@
+
+@QuarkusTest
+@QuarkusTestResource(KeycloakServer.class)
+public class UsersResourceTest {
```
@cemnura Can you please have a single test class (say `BearerTokenAuthenticationTest`) combining both admin and user resources tests - to have KC started and stopped once and avoid some instability.

cemnura replied:
@sberyozkin gladly. Resolving ASAP.

cemnura replied:
I combined the test classes into one class being `BearerTokenAuthenticationTest`.

cemnura commented:
I added a `@NativeImageTest` to satisfy PR checklist.

- [ ] has native tests (mvn clean verify -Pnative)

However, the native image test is failing. I have found the reason to be;

```java

    @ConfigProperty(name = "quarkus.oidc.auth-server-url")
    String oidc;
```
 
in the test class is set as `null` string during `mvn clean verify -Pnative` leading to the following error when trying to get access token;

```
  <body>RESTEASY003210: Could not find resource for full path: http://localhost:8081/null/protocol/openid-connect/token</body>
```

What would your recommendation be?

sberyozkin commented:
@cemnura Can you please use this version of `KeycloakServer`:
https://github.com/quarkusio/quarkus-quickstarts/blob/master/security-openid-connect-web-authentication-quickstart/src/test/java/org/acme/security/openid/connect/web/authentication/KeycloakServer.java
(Note it uses a fixed port), and then in the test use a keycloak realm URL with `8180`. (This is also done in `security-openid-connect-web-authentication-quickstart/.../application.properties`)

The way you inject it is interesting, can you please open a main Quarkus repository issue (something like `QuarkusTestResourceLifecycleManager properties can not be injected in the native image test`) and CC it to @stuartwdouglas ? This would be a nice feature...

Thanks

cemnura commented:
@sberyozkin I originally implemented a fixed port and reorganized TestContainer to use a random port according to previous recommendations. 
I would gladly reuse the;
https://github.com/quarkusio/quarkus-quickstarts/blob/master/security-openid-connect-web-authentication-quickstart/src/test/java/org/acme/security/openid/connect/web/authentication/KeycloakServer.java

I already tried it with native test and it definitely works.

 I will update the PR today.

> The way you inject it is interesting, can you please open a main Quarkus repository issue (something like QuarkusTestResourceLifecycleManager properties can not be injected in the native image test) and CC it to @stuartwdouglas ? This would be a nice feature...

Honestly I had no idea that it would even work I just tired it on a hence and it surprisingly worked. But it does not work with `@NativeImageTest`. I will create an issue regarding the usage of `@ConfigProperties` to be injected during `@QuarkusTest`

cemnura commented:
I removed the `@ConfigProperty` field in the test and made the TestContainer port fixed and made the test pass in both JVM and Native mode.

sberyozkin reviewed (approved):
Nice work @cemnura, thanks
